Implement Dereference pushdown for the Iceberg connector

alexjo2144 opened the pull request:
~This implementation only applies to tables using the ORC file format.~

~This is very much a WIP PR to get initial feedback~

Fixes https://github.com/trinodb/trino/issues/5179

TODO:
- Test with tables that don't have the Iceberg column id field

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -834,6 +844,120 @@ public void rollback()
+                icebergTableHandle.getSnapshotId(),
+                icebergTableHandle.getUnenforcedPredicate(),
+                icebergTableHandle.getEnforcedPredicate(),
+                Optional.of(projectedColumnsBuilder.build()));
```
That was based on the Hive implementation. Seems like the only place it's used is to avoid duplicating work if the same projections are given to the Metadata multiple times? https://github.com/trinodb/trino/pull/7360

It's also used for testing, but maybe we can skip it

losipiuk replied:
> Seems like the only place it's used is to avoid duplicating work if the same projections are given to the Metadata multiple times

This is actually important. The contract is that if call to `applyProjection` is a no-op should return Optional.empty()

alexjo2144 commented:
Finally had a chance to get back to this PR. I added the PageSource level projection adapter for if two column handles overlap like `a.b` and `a.b.c`. I'm still working on getting Parquet working but I should have some more time later this week.

alexjo2144 commented:
I added a second commit here with support for Parquet Iceberg tables. It was a little difference because the OrcReader accepts nested OrcColumns and knows how to read them back but parts of the ParquetReader column descriptions/types needed to be set up from the root of the column.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -77,6 +82,41 @@ public String getName()
+
+    public List<String> getFullPath()
+    {
+        return ImmutableList.<String>builder()
```
cache on field maybe

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -77,6 +82,41 @@ public String getName()
+            return getName();
+        }
+
+        return String.join(".", path) + "." + getName();
```
just make join on `getFullPath`

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -77,6 +82,41 @@ public String getName()
+    {
+        List<String> myPath = getFullPath();
+        List<String> otherPath = otherColumn.getFullPath();
+        return otherPath.size() >= myPath.size() && otherPath.subList(0, myPath.size()).equals(myPath);
```
`>=` or `>`? Is `x` parent of `x`

alexjo2144 replied:
Hmm, for the one place this is used it is important that x is a parent of x. Could probably inline this method though, I thought it was going to be more re-usable than it was.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +840,117 @@ public void rollback()
+
+        IcebergTableHandle icebergTableHandle = (IcebergTableHandle) handle;
+
+        String format = tableMetadataCache.get(icebergTableHandle.getSchemaTableName()).property("write.format.default", "ORC");
```
nit: you can do this as a first check in the method.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +840,111 @@ public void rollback()
+                    false));
+        }
+
+        Map<String, Assignment> newAssignments = new HashMap<>();
```
nit: ImmutableMap.builder()

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +840,111 @@ public void rollback()
+            ConnectorExpression expression = entry.getKey();
+            HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            IcebergColumnHandle projectedColumnHandle;
+            String projectedColumnName;
```
decalre on assignment.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +840,111 @@ public void rollback()
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+            projectedColumnName = projectedColumnHandle.getQualifiedName();
```
`projectedColumnName` will cotain dots. Not sure if that matters. What are we doing for Hive. Is that established pattern?

alexjo2144 replied:
> What are we doing for Hive?

Hive doesn't keep track of the projected columns in the same way. Instead of a path it keeps track of the base column and a list of integer deref indexes to drill down to the desired column. 

> Not sure if that matters

I'm not sure either. For the tests I ran, it didn't, but we could also switch them to a different symbol.

losipiuk replied:
No - if nothing breaks I assume this is fine.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSource.java:
```diff
@@ -96,6 +101,9 @@ public Page getNextPage()
                 return null;
             }
+            if (projectionsAdapter.isPresent()) {
+                dataPage = projectionsAdapter.get().adaptPage(dataPage);
```
check for `null` after adaptation?

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -176,10 +184,18 @@ public ConnectorPageSource createPageSource(
                 table.getUnenforcedPredicate());
 
-        return new IcebergPageSource(icebergColumns, partitionKeys, dataPageSource, session.getTimeZoneKey());
+        Optional<ReaderProjectionsAdapter> projectionsAdapter = Optional.empty();
+        if (dataPageSource.getReaderColumns().isPresent()) {
```
Use `dataPageSource.getReaderColumns().map()` instead.

losipiuk replied:
```java
        Optional<ReaderProjectionsAdapter> projectionsAdapter = dataPageSource.getReaderColumns().map(readerColumns ->         Optional<ReaderProjectionsAdapter> projectionsAdapter = dataPageSource.getReaderColumns().map(readerColumns -> 
                new ReaderProjectionsAdapter(
                        columns.stream().map(ColumnHandle.class::cast).collect(toImmutableList()),
                        readerColumns,
                        column -> ((IcebergColumnHandle) column).getType(),
                        IcebergPageSourceProvider::applyProjection));
                new ReaderProjectionsAdapter(
                        columns.stream().map(ColumnHandle.class::cast).collect(toImmutableList()),
                        readerColumns,
                        column -> ((IcebergColumnHandle) column).getType(),
                        IcebergPageSourceProvider::applyProjection));
```

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -176,10 +184,18 @@ public ConnectorPageSource createPageSource(
+        Optional<ReaderProjectionsAdapter> projectionsAdapter = Optional.empty();
+        if (dataPageSource.getReaderColumns().isPresent()) {
+            projectionsAdapter = Optional.of(new ReaderProjectionsAdapter(
+                    columns.stream().map(ColumnHandle.class::cast).collect(toImmutableList()),
```
just pass `columns` here and do defensive copy inside constructor if needed.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+
+        List<String> expectedDereferences = expectedColumn.getFullPath();
+        List<String> readerDereferences = readColumn.getFullPath();
+        checkState(expectedDereferences.subList(0, readerDereferences.size()).equals(readerDereferences),
```
`checkState(expectedColumn.isParentOf(readColumn))`

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> columns)
```
Add note to javadoc in `ReaderColumns` that iceberg uses it now. (or drop that only Hive uses it)

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+    {
+        ImmutableMap.Builder<Integer, OrcColumn> columnMap = ImmutableMap.builder();
+        columns.stream()
+                .filter(orcColumn -> orcColumn.getAttributes().containsKey(ORC_ICEBERG_ID_KEY))
```
Is it required that `ORC_ICEBERG_ID_KEY` will be present for each column? What happens if it is not?

alexjo2144 replied:
Thanks for the reminder. I think that case will likely either fail or return incorrect results. I was trying to find an existing test that covers that case, but I haven't been able to figure out when that property will be missing.

alexjo2144 replied:
I was looking through the Iceberg spec document and it seems to suggest that this should never happen. 

> Columns in Iceberg data files are selected by field id. The table schema’s column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be null.

https://iceberg.apache.org/spec/#column-projection

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+        // TODO: This is n^2 on the number of columns
+        for (Map.Entry<List<String>, Integer> existingMapping : mappedColumnIndices.entrySet()) {
+            List<String> existingPath = existingMapping.getKey();
+            if (column.getPath().size() >= existingPath.size() && column.getPath().subList(0, existingPath.size()).equals(existingPath)) {
```
Extract `public static boolean IcebergColumnHandle.isParent(List<String> maybeChild, List<String> maybeParent)` and use where needed. Also as an implementation of `ICH.isParentOf()`

losipiuk replied:
Or maybe what will be enough is to extract `ICH.isParentOf(List<String> path)` / `ICH.isChildOf(List<String> path` or both.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+
+    private static Optional<Integer> findExistingProjection(Map<List<String>, Integer> mappedColumnIndices, IcebergColumnHandle column)
+    {
+        // TODO: This is n^2 on the number of columns
```
how many columns can we have?

alexjo2144 replied:
Maybe not an unreasonable number, but we also run this code once per split and there might be a lot of splits

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+    private static Optional<Integer> findExistingProjection(Map<List<String>, Integer> mappedColumnIndices, IcebergColumnHandle column)
+    {
+        // TODO: This is n^2 on the number of columns
+        for (Map.Entry<List<String>, Integer> existingMapping : mappedColumnIndices.entrySet()) {
```
Instead iterating over existing mappings you can query `mappedColumnIndices` for each prefix of `column.getPath`. Then complexity will be O(n*max(path-length)) instead O(n^2)

alexjo2144 replied:
I like that. Thanks

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -269,11 +285,7 @@ private static ConnectorPageSource createOrcPageSource(
-                            identity()));
+            Map<Integer, OrcColumn> fileColumnsByIcebergId = mapIdsToOrcFileColumns(fileColumns);
             Map<String, OrcColumn> fileColumnsByName = null;
             if (fileColumnsByIcebergId.isEmpty()) {
```
this is preexistent; but check for `empty` is not obvious to me. Should we rather check if we are missing mapping for some columns?

alexjo2144 replied:
I think the assumption is that if the Id field exists for one column it must exist for all of them. I haven't seen a case where they don't exist yet.

losipiuk replied:
Can we add a checkstate that it holds? So the result map is either empty or the size matches fileColumns?

alexjo2144 replied:
Well, columns can be missing if they've been added since this data file was created.

alexjo2144 replied:
> if an old table is migrated to iceberg without rewriting data (which is going to be most common scenario initially), files won't contain ids at all.

Are there any tests for that situation? Would like to add a projected test case

alexjo2144 replied:
I added some migrated cases to the Spark compatibility tests. @phd3 are you familiar with the Iceberg `schema.name-mapping.default` property? Looks like a more robust way we could be mapping column names for migrated tables.

alexjo2144 replied:
If we had the full Iceberg schema here as suggested in https://github.com/trinodb/trino/pull/8754 I think we could get the mapping from `schema.findField(String name)`

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -283,10 +295,16 @@ private static ConnectorPageSource createOrcPageSource(
+                    .orElse(columns);
+            for (IcebergColumnHandle column : readColumns) {
                 OrcColumn orcColumn;
                 if (fileColumnsByIcebergId.isEmpty()) {
```
Is that possible at this point?

alexjo2144 replied:
It looks like it builds a backup mapping by name for the columns where the id is empty

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -283,10 +295,16 @@ private static ConnectorPageSource createOrcPageSource(
             List<ColumnAdaptation> columnAdaptations = new ArrayList<>(columns.size());
-            for (IcebergColumnHandle column : columns) {
+            List<IcebergColumnHandle> readColumns = columnProjections
+                    .map(readerColumns -> readerColumns.get().stream().map(IcebergColumnHandle.class::cast).collect(toList()))
```
nit: toImmutableList

alexjo2144 replied:
The generics of Optional.map don't seem to like having an ImmutableList and a List here. I guess the types don't work out?

losipiuk replied:
```java
            List<IcebergColumnHandle> readColumns = columnProjections
                    .map(readerColumns -> (List<IcebergColumnHandle>) readerColumns.get().stream().map(IcebergColumnHandle.class::cast).collect(toImmutableList()))
                    .orElse(columns);
```

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -283,10 +295,16 @@ private static ConnectorPageSource createOrcPageSource(
             Map<IcebergColumnHandle, Domain> effectivePredicateDomains = effectivePredicate.getDomains()
                     .orElseThrow(() -> new IllegalArgumentException("Effective predicate is none"));
+
+            Optional<ReaderColumns> columnProjections = projectColumns(columns);
```
Can we create `Optional<List<IcebergColumnHandle>>` instead of `Optional<ReaderColumns>`. It feels like only method we call on `ReaderColumns` is `get` anyway. And with that we could avoid this monster `.map(readerColumns -> readerColumns.get().stream().map(IcebergColumnHandle.class::cast).collect(toList()))` below :)

alexjo2144 replied:
The `ReaderProjectionsAdapter` expects `ReaderColumns` as an input, so if we're going to reuse that we need it.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +840,108 @@ public void rollback()
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
+            IcebergColumnHandle projectedColumnHandle = createProjectedColumnHandle(baseColumnHandle, projectedColumn.getDereferenceIndices(), expression.getType());
+            String projectedColumnName = projectedColumnHandle.getQualifiedName();
```
Is there a chance to get naming conflict here; get same `projectedColumnName` for different handles. It should not be possible if path elements and name cannot contain `dots`. It is a corner case, but still can you verify if that is possible?

alexjo2144 replied:
Looks like there's a validation for this in the Iceberg schema code:

```
trino:default> CREATE TABLE foo ("a.b" INT, a ROW (b INT));
Query 20210707_163341_00003_xkr2b failed: Invalid schema: multiple fields for name a.b: 1 and 3
org.apache.iceberg.exceptions.ValidationException: Invalid schema: multiple fields for name a.b: 1 and 3
```

losipiuk commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergOrcProjectionPushdown.java:
```diff
@@ -0,0 +1,228 @@
+import static java.lang.String.format;
+import static org.testng.Assert.assertTrue;
+
+public class TestIcebergOrcProjectionPushdown
```
Same for parquet?

losipiuk replied:
Or drop `Orc` from test name if not needed.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
```
nit: static import

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
```
Please ensure that we have a testcase when we are projecting two fields when one is child of the other (x.y and x.y.z). Maybe it is there - I did not check.

alexjo2144 replied:
Pretty sure this is what you're talking about: https://github.com/trinodb/trino/pull/8129/files#diff-8b18362088e5955f03b860a52d3ee8d3002f053c9e5ca76a888c27a272f9f0b7R1925

That one is effectively `select w, w.x.y.z`, nested a few levels deeper

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumnsForParquet(List<IcebergColumnHandle> columns)
```
What is the difference vs `projectColumnsForOrc`

alexjo2144 replied:
I meant to put this in the javadoc, but (IIUC) the Parquet reader expects columns to be referenced from the root of the file while Orc works fine referencing the nested columns directly. When I did Orc at first the ColumnHandle didn't need to keep track of the base column type, but for Parquet it does.

losipiuk commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
+        return columnMap.build();
+    }
+
+    private static Map<Integer, org.apache.parquet.schema.Type> mapIdsToParquetType(List<org.apache.parquet.schema.Type> fields)
```
I am too dumb for this. Please interleave the code with comments.

losipiuk replied:
Move it just below `projectColumnsForParquet` to keep parquet related code together.

losipiuk reviewed (commented):
Can we have a test which checks that less data is actually read due to dereference pushdown?

alexjo2144 commented:
> Can we have a test which checks that less data is actually read due to dereference pushdown?

Added :+1: 

AC, thanks @losipiuk

alexjo2144 commented:
@phd3 do you have time to give this another look over?

phd3 reviewed (commented):
still reviewing.

findepi commented:
@alexjo2144 there are conflicts. can you please rebase?

alexjo2144 commented:
Rebased

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -34,17 +36,27 @@
+            @JsonProperty("baseColumnIdentity") ColumnIdentity baseColumnIdentity,
+            @JsonProperty("baseType") Type baseType,
+            @JsonProperty("comment") Optional<String> comment,
+            @JsonProperty("path") List<String> path)
```
I feel we may be able to get away with storing less data here, especially duplicated columnIdentity and baseColumnIdentity. 

Ultimately, we may want to just start sending the serialized schema in IcebergTableHandle, and use `Schema` methods for field access. That'd save us from a lot of "reinventing the wheel" w.r.t. nested field access. (Not suggesting that this PR needs to be blocked on that.) cc @electrum 

For now, is it possible to simplify this if we use a map instead of a list in column identity? Where the ordering is required, we can use Trino type's fields. 

Also, I think it's better to store iceberg ids as much as possible (e.g. instead of strings in `path`), for robustness.

alexjo2144 replied:
Made a few changes: 
- Updated IcebergColumnHandle to just store the base column identity
- Updated the path to use int iceberg field ids instead of Strings
- Included a change to the `ColumnIdentity#children` representation to allow for quick lookup by index or by field id

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -356,6 +377,101 @@ private static ConnectorPageSource createOrcPageSource(
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumnsForOrc(List<IcebergColumnHandle> columns)
```
I just wasn't sure which would be better, the OrcReader seems to work either way but I can change it to project from the base columns. I had a todo for that https://github.com/trinodb/trino/pull/8129/files#diff-705548e3717786bbc076aba181f1eca081ec69200050c18100bfa959e05095dfR339

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -830,6 +841,112 @@ public void rollback()
+            HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn = entry.getValue();
+
+            // Create a new column handle
+            IcebergColumnHandle baseColumnHandle = (IcebergColumnHandle) assignments.get(projectedColumn.getVariable().getName());
```
Updated to include a `find` method

phd3 reviewed (commented):
ORC support lgtm

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergColumnHandle.java:
```diff
@@ -39,22 +39,25 @@
+                nestedColumnType,
+                nestedColumnType,
+                Optional.empty(),
+                ImmutableList.of());
```
nit: add test for a partial column

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1936,4 +1943,129 @@ protected TestTable createTableWithDefaultColumns()
+    {
+        String tableName = "projection_pushdown_after_rename";
+        try {
+            assertUpdate("CREATE TABLE " + tableName + "(id int, a ROW(b INT, c ROW (d int)))");
```
capitalize data types uniformly (also at other places)

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -1936,4 +1943,129 @@ protected TestTable createTableWithDefaultColumns()
+    }
+
+    @Test
+    public void testProjectionPushdownAfterRename()
```
let's also add a test for case-sensitive field names

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergProjectionPushdown.java:
```diff
@@ -0,0 +1,260 @@
+    }
+
+    @DataProvider(name = "format")
+    public static Object[][] formatProvider()
```
the plans are format-agnostic, I don't think this is required

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergProjectionPushdown.java:
```diff
@@ -0,0 +1,260 @@
+    private File metastoreDir;
+
+    @Override
+    protected LocalQueryRunner createLocalQueryRunner()
```
see `TestMetadataQueryOptimization#createQueryRunner` to reduce boilerplate

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergProjectionPushdown.java:
```diff
@@ -0,0 +1,260 @@
+                        filter(
+                                "x = BIGINT '5'",
+                                tableScan(
+                                        table -> {
```
compare unenforced constraint

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergProjectionPushdown.java:
```diff
@@ -0,0 +1,260 @@
+                                "y = BIGINT '2' AND (x =  cast((col1 + 3) as BIGINT))",
+                                tableScan(
+                                        table -> {
+                                            IcebergTableHandle icebergTableHandle = (IcebergTableHandle) table;
+                                            return icebergTableHandle.getProjectedColumns().equals(Optional.of(
+                                                    ImmutableSet.of(column1Handle, columnX, columnY)));
```
compare unenforced constraint

phd3 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergProjectionPushdown.java:
```diff
@@ -0,0 +1,260 @@
+                                                        tableScan(
+                                                                table -> true,
+                                                                TupleDomain.all(),
+                                                                ImmutableMap.of("x", equalTo(columnX), "expr_0", equalTo(column0Handle), "t_expr_1", equalTo(column1Handle))))),
```
compare unenforced constraint

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1001,6 +1012,171 @@ public void rollback()
+     * {@param assignments} with base="a" and path=[1, 2, 0].
+     * Note that the ProjectedColumnRepresentation uses index based dereferences but the IcebergColumnHandle#path uses Iceberg field ids.
+     */
+    public static Optional<String> find(Map<String, ColumnHandle> assignments, HiveApplyProjectionUtil.ProjectedColumnRepresentation projectedColumn)
```
this seems like a bit too much aerobics for what it's worth. Can we create a columnhandle and compare that instead before adding if that's simpler?

alexjo2144 commented:
Tests are failing due to https://github.com/trinodb/trino/pull/8822

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ColumnIdentity.java:
```diff
@@ -36,7 +37,8 @@
     private final TypeCategory typeCategory;
-    private final List<ColumnIdentity> children;
+    // LinkedHashMap is used to maintain the column ordering
+    private final LinkedHashMap<Integer, ColumnIdentity> children;
```
is it possible to avoid encoding order here and rely on `RowType` field ordering wherever required?

alexjo2144 replied:
I don't think so, not with still doing

> Also, I think it's better to store iceberg ids as much as possible (e.g. instead of strings in path), for robustness.

[from this comment](https://github.com/trinodb/trino/pull/8129#discussion_r677002820)

The issue I ran into while trying to use Types is that `ColumnIdentity` doesn't have `Types`, only `TypeCategory`'s, making it difficult to step down the dereference chain that way.

phd3 reviewed (commented):
Thanks for your efforts on this @alexjo2144! Overall this looks in good shape. Some comments, mostly minor. need to finish reviewing parquet changes.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:
```diff
@@ -1013,6 +1024,122 @@ public void rollback()
+                    false));
+        }
+
+        ImmutableMap.Builder<String, Assignment> newAssignments = ImmutableMap.builder();
+        ImmutableMap.Builder<ConnectorExpression, Variable> newVariablesBuilder = ImmutableMap.builder();
+        ImmutableSet.Builder<ColumnHandle> projectedColumnsBuilder = ImmutableSet.builder();
```
I had a `find` method but took it out, I'm still a bit confused on the reasoning for it.

From the other thread about it,
> IIRC duplicate avoidance in hive connector was intended to sanitize against the following case throwing duplicate-key error for ImmutableMap. Not sure if this is easily demonstrable in a query, but optimizer doesn't explicitly prevent it from happening.
> Assignments: <"a", CH(a)>, <"a_b", CH(a.b)>
> projections: ["a".b, "a_b"]
> if finding seems expensive, we can just verify that columnhandle is exactly the same and skip.

The Hive applyProjection still adds a new entry into `newAssignments`, `newVariablesBuilder`, and `projectedColumnsBuilder` even in the case where an existing column in the assignments is found. All it does is use the existing column handle and it's name rather than create a new one. I think I must be missing something because all I see that's different is skipping the construction of one, identical, column handle.

phd3 replied:
One thing to note here is that `assignments` is a map from String to ColumnHandle. The `String` is created from `Symbol#getName`. The symbol could've been created using the output of previous iteration of applyProjection.  say during one iteration, we return assignments `"a" -> CH(a), "a_b" -> CH(a.b)` from this method. Then the engine can create symbols `a_3` and `a_b_5` respectively.  Now, let's say for the next iteration, a projection like `"a_3".b` will produce variable named `a_b`, which is different from `a_b_5`. However, as long as `a_b_5` is also rewritten in the same manner, we wouldn't need to worry about having multiple variables for the same logical "subcolumn" in the returned assignments. (Which seems true here as well, so an explicit "find" is overly defensive, and not required.)

For the above testcase with the current code, we'll try to put the same string ("a_b") twice in `newAssignments`. and fail .So we shouldn't use immutable builder. It's okay for newVariablesBuilder as input connector expressions are different, and  also okay for projectedColumnsBuilder as it's a set and IcebergColumnHandle is created deterministically with properly defined equality. Does this make sense? If so, could you add a rule test? 

```
            Variable projectedColumnVariable = new Variable(projectedColumnName, expression.getType());
            Assignment newAssignment = new Assignment(projectedColumnName, projectedColumnHandle, expression.getType());
            newAssignments.put(projectedColumnName, newAssignment);

            newVariablesBuilder.put(expression, projectedColumnVariable);
            projectedColumnsBuilder.add(projectedColumnHandle);
```

alexjo2144 replied:
I took a stab at writing a rule test for this case, but the PlanMatcher didn't like it: https://github.com/trinodb/trino/pull/8129/files#diff-68db78292ccf93ffa9b1831d26475307dc02a4e1ad00b4c1913b26d65fb1f3ecR229-R254

cc: @findepi

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergColumnHandle.java:
```diff
@@ -32,24 +34,50 @@
+        return new IcebergColumnHandle(
+                createColumnIdentity(column),
+                type,
+                type,
```
This method was actually only ever getting called with an empty path parameter

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ColumnIdentity.java:
```diff
@@ -48,10 +50,14 @@ public ColumnIdentity(
                 "Children should be empty if and only if column type is primitive");
+        this.children = new LinkedHashMap<>(children.size());
+        for (ColumnIdentity child : children) {
+            this.children.put(child.getId(), child);
```
- validate no duplicates (`.put` doesn't override) and the put value isn't null (`rnn` on `child`)
- you should wrap the final map in the `unmodifiableMap`  to make it effectively immutable
- `ImmutableMap` also maintains insertion order and `ImmutableMap.builder` does all the above for you automatically


- relying on map order is quite error-prone; can we do better?

phd3 replied:
@findepi I think ultimately it'd the best if it's possible to use https://github.com/trinodb/trino/pull/8754, and get rid of this structure altogether. But I haven't tried the two PRs together to see if `Schema` method helps with simplifying this at all places. 

It looks like we need to the order when we translate a dereference into a subcolumn (and back).

One option is to keep Type structure as single source of truth at all places, but that'd require us to translate between (`RowType.Fields` ordering --> name --> ColumnIdentity's Id) and back. that seems a bit clunky.

Another option is to encode ordering in ColumnIdentity, but use something like 

```
    private final Map<Integer, ColumnIdentity> childrenById;
    private final List<Integer> idsForPositions;
```

and provide methods like `getChildAtPosition` and `getChildById`.  `getChildren()` method will return unordered map.

phd3 replied:
@findepi do you have any reservations against the second option above? I think it'd be atleast less error-prone than using `ImmutableMap` or `linkedinhashmap`. (but slightly clunkier during creation)

findepi replied:
> Another option is to encode ordering in ColumnIdentity, but use something like
> 
> ```
>     private final Map<Integer, ColumnIdentity> childrenById;
>     private final List<Integer> idsForPositions;
> ```

works for me.

or perhaps simpler:
```java
private final List<ColumnIdentity> children;
private final Map<Integer, ColumnIdentity> childrenById;
```

> I think it'd be atleast less error-prone than using `ImmutableMap` or `linkedinhashmap`.

I realized we do not return the map in any getter, so the current code (using ImmutableMap) works for me too.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -534,6 +666,68 @@ private static ConnectorPageSource createParquetPageSource(
+                                .stream()
+                                .collect(toMap(
+                                        Map.Entry::getKey,
+                                        // Preserve the parent type details so the schema can be constructed from the root of the type tree
+                                        entry -> new GroupType(field.getRepetition(), field.getName(), ImmutableList.of(entry.getValue())))));
```
I don't understand what it does, but the final result feels hacky: in the returned map under some ID key K we don't have the parquet Type with Id K.

It could be that https://github.com/trinodb/trino/pull/9104 solves this more elegantly, by changing `ParquetColumnIOConverter`.

alexjo2144 replied:
This method is actually unnecessary now that we're projecting based off of the base column all the time. I'll roll this change back.

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ColumnIdentity.java:
```diff
@@ -54,10 +55,11 @@ public ColumnIdentity(
                 children.isEmpty() == (typeCategory == PRIMITIVE),
                 "Children should be empty if and only if column type is primitive");
-        this.children = new LinkedHashMap<>(children.size());
+        ImmutableMap.Builder<Integer, ColumnIdentity> childrenBuilder = ImmutableMap.builder();
```
update `// LinkedHashMap is used to maintain ..` above

alexjo2144 commented:
@findepi @phd3 I rebased to pull in Piotr's changes, some of the new tests found some bugs but they're passing now. One was a mixup between the Orc id and the Iceberg field id, that's fixed in the last fixup. The other was a bug in the Orc reader, using the wrong column name in the case where the RowType name was different from the Orc schema. That's in a prep commit.

alexjo2144 commented:
Rebased and resolved conflicts. I moved the Rule Test I started writing over to a gist since it wasn't running properly: https://gist.github.com/alexjo2144/7a897056d6b3c6f0bd876470948aca00

findepi commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -563,6 +658,81 @@ private static ConnectorPageSource createParquetPageSource(
+                return fullyProjectedLayout();
+            }
+
+            Map<Integer, List<List<Integer>>> dereferencesByField = fieldIdDereferences.stream().collect(
+                    Collectors.groupingBy(
+                            sequence -> sequence.get(0),
+                            mapping(sequence -> sequence.subList(1, sequence.size()), toList())));
```
It seems we're doing tree traversal here (effectively). 
wonder whether we can improve representation of dereferences so that we don't need to copy all the elements when recursing (OK to address in follow-up)

findepi replied:
It seems we're doing tree traversal here (effectively). 
wonder whether we can improve representation of dereferences so that we don't need to copy all the elements when recursing (OK to address in follow-up)

sopel39 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -319,14 +347,30 @@ private static ConnectorPageSource createOrcPageSource(
-                    Domain domain = effectivePredicateDomains.get(column);
-                    if (domain != null) {
-                        predicateBuilder.addColumn(orcColumn.getColumnId(), domain);
+                    for (Map.Entry<IcebergColumnHandle, Domain> domainEntry : effectivePredicateDomains.entrySet()) {
```
Does the same issue apply to Hive connector?

alexjo2144 replied:
In Hive, for each read column we're first filtering out the domains from the `effectivePredicate` that matter to that column: https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/orc/OrcPageSourceFactory.java#L345

It's about the same, a linear filter over the predicate column for each base column.

@findepi we could simplify the code here a little now that we're only ever projecting base columns. I wrote `covers` to work even if both of the columns were projected subfields (a.b covers a.b.c), but now we know that we're only reading base columns, we could just compare `IcebergColumnHandle#getBaseColumnIdentity`

findepi replied:
>  we could just compare IcebergColumnHandle#getBaseColumnIdentity

that would allow us for a faster search

however, is there anything guaranteeing predicated columns are only base columns?
if we push `struct.field` projection down into Iceberg connector,
then we can subsequently push `struct.field > 0` filter into it as well, right?

alexjo2144 replied:
The predicate columns are not guaranteed to be base columns, but the read column is. So we'd be comparing `readColumn.getColumnIdentity().equals(predicateColumn.getBaseColumnIdentity())` to filter out predicates that belong to that read column.

I'm adding some `verify`s to assert that the read column is always a base column. It's not super clear that's the case right now, but it is.

alexjo2144 commented:
Had to rebase to resolve merge conflicts but the last set of comments is in the last fixup commit. Thanks

alexjo2144 commented:
Test failures are in `trino-jdbc`, unrelated

findepi commented:
> TODO:
> 
> * Test with tables that don't have the Iceberg column id field

please make sure there is an issue for this

> Test failures are in `trino-jdbc`, unrelated

please make sure there is an issue for this

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -481,7 +567,12 @@ private static ConnectorPageSource createParquetPageSource(
+                    .map(readerColumns -> (List<IcebergColumnHandle>) readerColumns.get().stream().map(IcebergColumnHandle.class::cast).collect(toImmutableList()))
+                    .orElse(regularColumns);
+
+            List<org.apache.parquet.schema.Type> parquetFields = readColumns.stream()
```
are we intentionally not pushing predicates into parquet reader? (line 591)

alexjo2144 replied:
Probably just because it's relatively new. The change allowing for passing a Predicate to ParquetReader was just merged in August. If it's useful here, I'd rather do it in followup, since it's related to all the column index changes to the parquet reader and I don't really understand those yet. https://github.com/trinodb/trino/commit/6eb42f2aeca9a07a53249ae4bd56c13c8f002335

phd3 replied:
I was mainly thinking of the pushdown in rowgroups (ORC stripe-equivalent, even before newly added code). but look like it's missing in hive connector too. cc @JamesRTaylor

findepi replied:
i am not sure i understand the conclusion. should we have a follow-up issue for this?

phd3 replied:
[exclusion](https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java#L201) of nested predicates was added initially when only ORC support was added for dereference pushdown. However, seems like it didn't change in #3396 while adding parquet support. I'm not sure if that was intentional. but you're right that this deserves a separate issue. #9928

For iceberg connector, I don't see corresponding changes like #3396 here for [propagating projected layout](https://github.com/trinodb/trino/pull/8129#discussion_r745227296), so assumed that we're tackling it separately.

phd3 replied:
#9931

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -319,14 +351,30 @@ private static ConnectorPageSource createOrcPageSource(
                         readType = column.getType();
                     }
 
+                    List<List<Integer>> fieldIdProjections = fileColumnsByIcebergId.isEmpty() ?
```
seems like it'll break?
```
    @Test
    public void testNestedUuid()
    {
        String tableName = "test_nested_uuid";
        assertUpdate(format("CREATE TABLE %s (x row(a uuid, b bigint), y bigint)", tableName));

        assertUpdate(format("INSERT INTO %s VALUES (row(UUID '406caec7-68b9-4778-81b2-a12ece70c8b1', 1), 2)", tableName), 1);
        assertQuery(format("SELECT x.a FROM %s", tableName), "SELECT CAST('406caec7-68b9-4778-81b2-a12ece70c8b1' AS UUID)");
    }
```

alexjo2144 replied:
Looks like UUIDs within array/map/row types doesn't work for ORC. I'll add in a fix for those. cc: @findepi

phd3 replied:
ah, it's pre-existing then. Let me resolve this. Generally I'd expect someone to mostly use this type at top level anyway.

alexjo2144 replied:
https://github.com/trinodb/trino/pull/9930

phd3 reviewed (changes requested):
thanks for your work on this @alexjo2144.

With Iceberg, pushdown on metadata side itself gives us huge benefits because of split generation time pruning. However, I think there're some missing pieces on pagesource side as mentioned in comments. Please let me know if there's an issue with my understanding.

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -319,14 +351,30 @@ private static ConnectorPageSource createOrcPageSource(
+                    List<List<Integer>> fieldIdProjections = fileColumnsByIcebergId.isEmpty() ?
+                            projectionsByName.get(column.getBaseColumnIdentity().getName()) :
+                            projectionsByFieldId.get(column.getId());
+                    ProjectedLayout projectedLayout = IcebergOrcProjectedLayout.createProjectedLayout(orcColumn, fieldIdProjections);
```
This is probably wrong but I'd like to address it in a follow up PR under the other issue: https://github.com/trinodb/trino/issues/9843

Based on the comments from Ryan in there, all of the name based code we have here is probably not correct in the first place.

phd3 replied:
thanks. My only concern is that this will cause regression for users who haven't added the mapping field. Should we at least deprecate first? Or do we just say that dereference pushdown can be only used when IDs are present. i.e. projection pushdown flag serves as deprecation flag for name/id too? cc @findepi

findepi replied:
The current code handling migrated Hive files (without iceberg IDs) is wrong, produces incorrect results and needs to be rewritten anyway. I don't think it's practical to hold on dereference pushdown until that happens. Rather, I would prefer to temporarily disable handling of migrated files. Would that be better?

phd3 replied:
> I don't think it's practical to hold on dereference pushdown until that happens.

yeah, I don't mean to suggest that at all. 

> I would prefer to temporarily disable handling of migrated files.

My concern is that this will break queries for folks in the process of hive->migration, and can prevent Trino upgrade.

I see your point as per #9843, that name-based mapping is wrong w.r.t. Iceberg spec, but I'd expect this to be very common for Trino users in absence of iceberg ids. So IMO it'll be better to keep supporting it until we've  support for the "correct" mapping using `schema.name-mapping.default` - if it's a relatively low overhead (implementing a couple methods).

phd3 replied:
cc @electrum

phd3 replied:
Let me attempt at consolidating the discussion:

* This only concerns Iceberg tables with files that do not have Iceberg IDs. 

* For tables without “schema.name-mapping.default” property:
  * Before this PR: Trino uses simple name-based mapping. This also seems consistent with Iceberg readers. (e.g. https://github.com/apache/iceberg/pull/580)
  * After this PR: dereference queries will fail on these tables. This'd be a regression in the next release. 

* For tables with “schema.name-mapping.default” property:
  * Before this PR: Trino uses simple name-based mapping. This is incorrect.
  * After this PR: dereference queries will fail on these tables. This is better than silently returning incorrect results, however it is not exhaustive.

* To avoid regression in the first case, we may need to add more changes in the same release. If proper handling of “schema.name-mapping.default” takes more time/effort, we may want to at least restore name-based mapping for the case when  “schema.name-mapping.default” is absent.

alexjo2144 replied:
That is a good summary, thanks. I spent some time this afternoon looking at using `schema.name-mapping.default`. It looks pretty straightforward.

alexjo2144 replied:
I put up a change to use the default name mapping here: https://github.com/trinodb/trino/pull/9959

It's based on this PR

alexjo2144 replied:
> For tables without “schema.name-mapping.default” property:

> Before this PR: Trino uses simple name-based mapping. This also seems consistent with Iceberg readers. (e.g. apache/iceberg#580)

As far as consistency with Spark, that behavior changes depends on the reader, Spark using Iceberg v0.12 has different results reading ORC vs Parquet. According to Ryan https://github.com/trinodb/trino/issues/9843#issuecomment-970514001 the values should be `null`

phd3 replied:
thanks for following up on this!

alexjo2144 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -319,14 +351,30 @@ private static ConnectorPageSource createOrcPageSource(
+                        IcebergColumnHandle predicateColumn = domainEntry.getKey();
+                        OrcColumn predicateOrcColumn;
+                        if (fileColumnsByIcebergId.isEmpty()) {
+                            predicateOrcColumn = fileColumnsByName.get(predicateColumn.getName().toLowerCase(ENGLISH));
```
Same as https://github.com/trinodb/trino/pull/8129#discussion_r745956411

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableHandle.java:
```diff
@@ -39,21 +41,25 @@
     // Filter guaranteed to be enforced by Iceberg connector
     private final TupleDomain<IcebergColumnHandle> enforcedPredicate;
 
+    private final Set<IcebergColumnHandle> projectedColumns;
```
ah, okay. it seems a bit confusing to me, but since Piotr suggested otherwise, let's go with that. resolving.

phd3 reviewed (approved):
I think we can merge this one PR with ORC improvements once https://github.com/trinodb/trino/pull/8129#discussion_r745218223 is resolved, and handle Parquet stuff separately. Can you please squash commits?

findepi commented:
CI https://github.com/trinodb/trino/issues/8611

alexjo2144 commented on lib/trino-orc/src/main/java/io/trino/orc/OrcReader.java:
```diff
@@ -432,29 +432,37 @@ static void validateFile(
+        ProjectedLayout getFieldLayout(OrcColumn orcColumn);
+    }
+
+    public static ProjectedLayout fullyProjectedLayout()
```
Added. For what it's worth, this pattern is already what we're using, it's just that the infinite recursion is done over a few methods.

https://github.com/trinodb/trino/blob/master/lib/trino-orc/src/main/java/io/trino/orc/OrcReader.java#L455
 `fullyProjectedLayout()` -> `getFieldLayout()` -> `fullyProjectedLayout` ... etc

alexjo2144 commented:
Created a follow up Issue: https://github.com/trinodb/trino/issues/9931

phd3 commented on plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:
```diff
@@ -481,7 +567,12 @@ private static ConnectorPageSource createParquetPageSource(
             boolean mapByName = parquetIdToField.isEmpty();
 
-            List<org.apache.parquet.schema.Type> parquetFields = regularColumns.stream()
+            Optional<ReaderColumns> columnProjections = projectColumns(regularColumns);
```
#9931

phd3 reviewed (approved):
Thanks for your work on this @alexjo2144 ! I'll merge this unless someone else has further comments.

phd3 commented:
Merged, thanks!
